### ISSUE TITLE
When saving menus the automated cache purge is triggered too many times

### ISSUE THREAD

**Issue by alewolf** [cloudflare/Cloudflare-WordPress#275]
We ran into an issues where WordPress would time out each time we save a menu. 

Saving the menu is done within seconds when the automatic cache purge is turned off. But with the automatic cache purge turned on the operation takes too long and returns a 502. 

We found when saving a menu, dozens (if not hundreds) of cache purge requests are triggered by the plugin. 

The workaround is obvious. Before saving a menu the automatic cache purge needs to be turned off. 

But I believe this should be fixed in the plugin itself. 

Since you can't be sure that these multi trigger events don't only happen on menu saves I would suggest something similar as the following: 

Queue the cache purges within WordPress: For instance trigger one cache purge immediately and queue one to be executed in a minute from now. Drop all other purge cache requests as long as there is a queued purge requests waiting to be sent to Cloudflare.

**Comment by rmens** [comment]
I could reproduce this with the latest version of the plug-in. There seems to be a limited number of purge requests that can be sent to the API. Making a queue would be a great idea (https://github.com/cloudflare/Cloudflare-WordPress/issues/266)

**Comment by gonzomir** [comment]
I ran into this issue too, causing CloudFlare to timeout waiting for the origin server to update the menu.

One of the things the plugin must do is check if the saved post type is public before anything else. There's no point in purging the cache for things that are not public, hence not cached. This applies to many other things, like WooCommerce orders or order items for example.

Then, there should be separate handling of saving menus. There may be menus that are not displayed anywhere on the site, and no cache needs to be purged, the primary navigation obviously calls for purging all pages, etc.

**Comment by mayaliny** [comment]
I think we ran in the same problem today.. basically when we were saving a wordpress menu, we keep getting timeouts cause something like 110 purge requests were sent to CF. Every purge took like 0.88sec, so out total was around 50seconds.

We raised the apache timeout to 60 seconds, but there's no point at all to purge pages individually. When a Wordpress menu that is present in every page is changed ALL the page cache must be purged OR requests have to be managed in groups OR the Wordpress part has not to wait for the CF response.

**Comment by sejoker** [comment]
I agree we should execute Cache purge operations as a background jobs. We are open for contributions.

**Comment by sejoker** [comment]
The manual steps  to workaround the issue:
- Disable APO /"Auto Purge Content On Update".
- Save menus
- Purge everything
- Enable  APO /"Auto Purge Content On Update".

**Comment by rmens** [comment]
This is a valid reproducible issue. It should not be closed.
